### PR TITLE
News List

### PR DESCRIPTION
**Pull request by Woozl**
> [!IMPORTANT]
> As the tag endpoint is still finalized, you currently need to use a plugin like [ModHeader](https://chromewebstore.google.com/detail/modheader-modify-http-hea/idgpnmonknjnojddfkpgkljpfnnfcklj?pli=1) to allow the client to access `api.renci.org/graphql`, by providing a Strapi API token in the `Authorization` header:
> ```
> "Authorization": "Bearer <API_TOKEN>"
> ```
> I think we'll want to wait until that endpoint is finalized before merging this, but we can go ahead testing this UI in the meantime.

features:
- state in query params
- tag + article fetches on client side
- skeleton loaders for tag + article fetches
- error message component (_not currently implemented for tag fetch until endpoint is ready_)
- url+query param caching with SWR (_not currently implemented for tag fetch until endpoint is ready_)
- pagination
- clickable tags in article preview
- custom tags for arbitrary text to search the article title and slugs
- custom tags matches are highlighted in article preview
- filter drawer when in mobile view
- message when there are no article results
- sorted tags in autocomplete (AC) and filter list
- tags are colored according to type
- filtered searching with highlighting when searching through tags in AC
- clear all button to remove all filters

Waterfall skeleton loaders, query params, custom search with highlighted titles, pagination:

https://github.com/mbwatson/renci-dot-org/assets/16181779/e7daf5c8-c59c-40e1-9ec2-10a66c5c6855

Clickable tags, using browser back button to update to move state back, going back has no loader as the request has been cached:

https://github.com/mbwatson/renci-dot-org/assets/16181779/a99de5e5-9edf-48dc-be65-6558589b7a49

Mobile view, drawer opened by button with badge showing number of active filters, "no results" text:

https://github.com/mbwatson/renci-dot-org/assets/16181779/2d59ab38-f370-4d42-b758-534ad348932a

### CODE REVIEW

File: components/news/news-or-feature-toggle.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { Button, ButtonGroup } from "@mui/material";
+
+export const NewsOrFeatureToggle = ({ newsOrFeature, setNewsOrFeature }) => (
```
Comment (mbwatson) [inline components/news/news-or-feature-toggle.js]: should this be FeatureOrBlogToggle or BlogOrFeatureToggle? perhaps better would be to not bake the options into the function name. i'd suggest something a bit more generic, like ArticleTypeToggle or something similar.

File: components/news/news-or-feature-toggle.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  <ButtonGroup sx={{ display: 'flex', '--br': '4px', px: '8px' }}>
```
Comment (mbwatson) [inline components/news/news-or-feature-toggle.js]: Button descendants will inherit the size prop from ButtonGroup, so i think we can clean up some repetition there.

In addition, I think that many times it's cleaner to pull all the styling to the top level of component definitions. this isn't a mandate, but something such as this feels a bit easier to parse and maintain:
```
  <ButtonGroup
    size="small"
    sx={{
      display: 'flex', '--br': '4px', px: '8px',
      '.MuiButton-root:first-child': { flex: 1, borderRadius: 'var(--br) 0 0 var(--br)' },
      '.MuiButton-root': { flex: 1 },
      '.MuiButton-root:last-child': { flex: 1, borderRadius: '0 var(--br) var(--br) 0' },
    }}
  >
    <Button
      onClick={() => { setNewsOrFeature(null); }}
      variant={newsOrFeature === null ? "contained" : "outlined"}
    >
      All
    </Button>
    <Button
      onClick={() => { setNewsOrFeature("news"); }}
      variant={newsOrFeature === "news" ? "contained" : "outlined"}
    >
      News
    </Button>
    <Button
      onClick={() => { setNewsOrFeature("feature"); }}
      variant={newsOrFeature === "feature" ? "contained" : "outlined"}
    >
      Feature
    </Button>
  </ButtonGroup>
```

File: lib/strapi/newsGraphQL.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+*   researchGroups: Array<{ slug: string, name: string, numOfPosts: number }>
+*   collaborations: Array<{ slug: string, name: string, numOfPosts: number }>
+*   projects: Array<{ slug: string, name: string, numOfPosts: number }>
+*   organizations: Array<{ slug: string, name: string, numOfPosts: number }>
+*   people: Array<{ slug: string, name: string, numOfPosts: number }>
+*   postTags: Array<{ name: string, numOfPosts: number }>
+* }> }
+*/
+export const fetchTags = async () => {
+ const frag = (type = 'generic') => `(pagination: {limit: 500}) {
```
Comment (mbwatson) [inline lib/strapi/newsGraphQL.js]: :sunglasses: clever

File: components/news/tag.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
+import PropTypes from "prop-types";
+import { Chip } from "@mui/material";
+import { CloseRounded } from "@mui/icons-material";
+
+const TYPES = {
```
Comment (mbwatson) [inline components/news/tag.js:5]: at some point, i suspect we'll revisit these colors

File: lib/strapi/fetchStrapiGraphQL.js [hunk 5]
```diff
@@ -1,10 +1,11 @@
-export const fetchStrapiGraphQL = async (query, preview = false) => {
+export const fetchStrapiGraphQL = async (query, preview = false, signal = undefined) => {
```
Comment (mbwatson) [inline lib/strapi/fetchStrapiGraphQL.js:1]: :+1:

**Review by mbwatson — APPROVED** [review]
this is awesome, David. this has the features we're after and the implementation is solid. if there was a place to call out that has the most room for improvement, i'd say it's news/index.js, purely because there's a lof of functionality baked into the page component. i think long-term maintenance would be easier if the business logic for the query params/tags was extracted to its own hook, like useQueryTags or similar.

in any event, this is all great :+1:. i left a few nitpicky comments, but nothing that absolutely needs to be changed before merging. i know this feature will continue to see activity as other things fall into place.

File: components/news/news-or-feature-toggle.js (hunk 2 above)
Comment (mbwatson) [reply components/news/news-or-feature-toggle.js]: there were other opportunities to make this same comment, but i didn't note it more than this one time.

**Review by scrummish — APPROVED** [review]

**Pull request merged**